cpp/examples/arrow/row-wise-conversion-example.cc

```diff
@@ -0,0 +1,190 @@
+  std::vector<std::shared_ptr<arrow::Field>> schema_vector = {
+      arrow::field("id", arrow::int64()), arrow::field("cost", arrow::float64()),
+      arrow::field("cost_components", arrow::list(arrow::float64()))};
+  auto expected_schema = std::make_shared<arrow::Schema>(schema_vector);
+
+  if (!expected_schema->Equals(*table->schema())) {
+    // The table doesn't have the expected schema thus we cannot directly
+    // convert it to our target representation.
+    return arrow::Status::Invalid("Schemas are not matching!");
+  }
+
+  // As we have ensured that the table has the expected structure, we can unpack the
+  // underlying arrays. For the primitive columns `id` and `cost` we can use the high
+  // level functions to get the values whereas for the nested column
+  // `cost_components` we need to access the C-pointer to the data to copy its
+  // contents into the resulting `std::vector<double>`. Here we need to be care to
+  // also add the offset to the pointer. This offset is needed to enable zero-copy
+  // slicing operations. While this could be adjusted automatically for double
+  // arrays, this cannot be done for the accompanying bitmap as often the slicing
+  // border would be inside a byte.
+
+  auto ids =
+      std::static_pointer_cast<arrow::Int64Array>(table->column(0)->data()->chunk(0));
+  auto costs =
+      std::static_pointer_cast<arrow::DoubleArray>(table->column(1)->data()->chunk(0));
+  auto cost_components =
+      std::static_pointer_cast<arrow::ListArray>(table->column(2)->data()->chunk(0));
+  auto cost_components_values =
+      std::static_pointer_cast<arrow::DoubleArray>(cost_components->values());
+  // To enable zero-copy slices, the native values pointer might need to account
+  // for this slicing offset. This is not needed for the higher level functions
+  // like Value(…) that already account for this offset internally.
+  const double* ccv_ptr = cost_components_values->data()->GetValues<double>(1);
+
+  for (int64_t i = 0; i < table->num_rows(); i++) {
+    // Another simplification in this example is that we assume that there are
+    // no null entries, e.g. each row is fill with valid values.
+    int64_t id = ids->Value(i);
+    double cost = costs->Value(i);
+    const double* first = ccv_ptr + cost_components->value_offset(i);
+    const double* last = ccv_ptr + cost_components->value_offset(i + 1);
+    std::vector<double> components_vec(first, last);
+    rows->push_back({id, cost, components_vec});
+  }
+
+  return arrow::Status::OK();
+}
+
+#define EXIT_ON_FAILURE(expr)                      \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      return EXIT_FAILURE;                         \
+    }                                              \
+  } while (0);
```

Comment:
Do we want to add an `ARROW_ABORT_NOT_OK`?

Reply:
OK. It is not necessary to resolve this here, but it would be useful for the library to come with various Status-handling macros built-in to accommodate different needs, so users don't have to implement their own.